deals/module/store/store.go

```diff
@@ -0,0 +1,376 @@
+		}
+		ret = append(ret, dr)
+	}
+	return ret, nil
+}
+
+// GetFinalStorageDeals returns all final storage deal records.
+func (s *Store) GetFinalStorageDeals() ([]deals.StorageDealRecord, error) {
+	s.lock.Lock()
+	defer s.lock.Unlock()
+
+	q := query.Query{Prefix: dsBaseStorageFinal.String()}
+	res, err := s.ds.Query(q)
+	if err != nil {
+		return nil, fmt.Errorf("executing query: %s", err)
+	}
+	defer func() {
+		if err := res.Close(); err != nil {
+			log.Errorf("closing query result: %s", err)
+		}
+	}()
+
+	var ret []deals.StorageDealRecord
+	for r := range res.Next() {
+		if r.Error != nil {
+			return nil, fmt.Errorf("iter next: %s", r.Error)
+		}
+		var dealRecord deals.StorageDealRecord
+		if err := json.Unmarshal(r.Value, &dealRecord); err != nil {
+			return nil, fmt.Errorf("unmarshaling query result: %s", err)
+		}
+		ret = append(ret, dealRecord)
+	}
+
+	return ret, nil
+}
+
+// PutRetrieval saves a retrieval deal record.
+func (s *Store) PutRetrieval(rr deals.RetrievalDealRecord) error {
+	s.lock.Lock()
+	defer s.lock.Unlock()
+
+	txn, err := s.ds.NewTransaction(false)
+	if err != nil {
+		return fmt.Errorf("creating transaction: %s", err)
+	}
+	defer txn.Discard()
+
+	rr.ID = retrievalID(rr)
+	rr.UpdatedAt = time.Now().UnixNano()
+	buf, err := json.Marshal(rr)
+	if err != nil {
+		return fmt.Errorf("marshaling RetrievalRecord: %s", err)
+	}
+	key := makeRetrievalKey(rr)
+	if err := txn.Put(key, buf); err != nil {
+		return fmt.Errorf("put RetrievalRecord: %s", err)
+	}
+
+	updatedAtIndexKey := makeRetrievalUpdatedAtIndexKey(rr.UpdatedAt)
+	if err := txn.Put(updatedAtIndexKey, key.Bytes()); err != nil {
+		return fmt.Errorf("saving updated-at index: %s", err)
+	}
+
+	if err := txn.Commit(); err != nil {
+		return fmt.Errorf("committing transaction: %s", err)
+	}
+
+	return nil
+}
+
+// GetRetrievals returns all retrieval deal records.
+func (s *Store) GetRetrievals() ([]deals.RetrievalDealRecord, error) {
+	s.lock.Lock()
+	defer s.lock.Unlock()
+
+	q := query.Query{Prefix: dsBaseRetrieval.String()}
+	res, err := s.ds.Query(q)
+	if err != nil {
+		return nil, fmt.Errorf("executing query: %s", err)
+	}
+	defer func() {
+		if err := res.Close(); err != nil {
+			log.Errorf("closing query result: %s", err)
+		}
+	}()
+
+	var ret []deals.RetrievalDealRecord
+	for r := range res.Next() {
+		if r.Error != nil {
+			return nil, fmt.Errorf("iter next: %s", r.Error)
+		}
+		var rr deals.RetrievalDealRecord
+		if err := json.Unmarshal(r.Value, &rr); err != nil {
+			return nil, fmt.Errorf("unmarshaling query result: %s", err)
+		}
+		ret = append(ret, rr)
+	}
+	return ret, nil
+}
+
+// GetUpdatedStorageDealRecordsSince returns all the storage deal records that got created or updated
+// since sinceNano.
+func (s *Store) GetUpdatedStorageDealRecordsSince(sinceNano int64, limit int) ([]deals.StorageDealRecord, error) {
+	q := query.Query{
+		Prefix: dsStorageUpdatedAtIdx.String(),
+	}
```

Comment:
Whenever we can dedicate time to adding app metrics and having good dashboards, we can detect when queries start taking more time than reasonable. At that point, we can think about pruning or doing something.

I think this kind of metrics are important since we're walking in the dark here about how bad this might be.
I guess we eventually need to push this as a prioritization.